model/infinispan/src/main/java/org/keycloak/spi/infinispan/impl/embedded/DefaultCacheEmbeddedConfigProviderFactory.java

```diff
@@ -0,0 +1,250 @@
+
+/**
+ * The default implementation of {@link CacheEmbeddedConfigProviderFactory}.
+ * <p>
+ * It builds a {@link ConfigurationBuilderHolder} based on the Keycloak configuration.
+ * <p>
+ * Advanced users may extend this class and overwrite the method {@link #createConfiguration(KeycloakSessionFactory)}.
+ * They have access to the {@link ConfigurationBuilderHolder}, and they can modify it as needed for their custom
+ * providers.
+ */
+public class DefaultCacheEmbeddedConfigProviderFactory implements CacheEmbeddedConfigProviderFactory, CacheEmbeddedConfigProvider {
+
+    private static final Logger logger = Logger.getLogger(MethodHandles.lookup().lookupClass());
+
+    public static final String PROVIDER_ID = "default";
+
+    // Configuration
+    public static final String CONFIG = "configFile";
+    private static final String METRICS = "metricsEnabled";
+    private static final String HISTOGRAMS = "metricsHistogramsEnabled";
+    public static final String STACK = "stack";
+    public static final String NODE_NAME = "nodeName";
+    public static final String SITE_NAME = "siteName";
+
+    private volatile ConfigurationBuilderHolder builderHolder;
+    private volatile Config.Scope keycloakConfig;
+
+    @Override
+    public CacheEmbeddedConfigProvider create(KeycloakSession session) {
+        lazyInit(session.getKeycloakSessionFactory());
+        return this;
+    }
+
+    @Override
+    public void init(Config.Scope config) {
+        this.keycloakConfig = config;
+    }
+
+    @Override
+    public void postInit(KeycloakSessionFactory factory) {
+        lazyInit(factory);
+    }
+
+    @Override
+    public ConfigurationBuilderHolder configuration() {
+        return builderHolder;
+    }
+
+    @Override
+    public void close() {
+        //no-op
+    }
+
+    @Override
+    public String getId() {
+        return PROVIDER_ID;
+    }
+
+    @Override
+    public List<ProviderConfigProperty> getConfigMetadata() {
+        var builder = ProviderConfigurationBuilder.create();
+        Util.copyFromOption(builder, CONFIG, "file", ProviderConfigProperty.STRING_TYPE, CachingOptions.CACHE_CONFIG_FILE, false);
+        Util.copyFromOption(builder, HISTOGRAMS, "enabled", ProviderConfigProperty.BOOLEAN_TYPE, CachingOptions.CACHE_METRICS_HISTOGRAMS_ENABLED, false);
+        Util.copyFromOption(builder, METRICS, "enabled", ProviderConfigProperty.BOOLEAN_TYPE, MetricsOptions.INFINISPAN_METRICS_ENABLED, false);
+        Stream.concat(Arrays.stream(LOCAL_CACHE_NAMES), Arrays.stream(CLUSTERED_MAX_COUNT_CACHES))
+                .forEach(name -> Util.copyFromOption(builder, CacheConfigurator.maxCountConfigKey(name), "max-count", ProviderConfigProperty.INTEGER_TYPE, CachingOptions.maxCountOption(name), false));
+        createTopologyProperties(builder);
+        return builder.build();
+    }
+
+    @Override
+    public Set<Class<? extends Provider>> dependsOn() {
+        return Set.of(JGroupsCertificateProvider.class);
```

Comment:
Not related to these changes per se, but I noticed that `RemoteLoadBalancerCheckProviderFactory` depends on `InfinispanConnectionProvider` but does not have a `dependsOn` implementation. Is there a reason for this, or should I create an issue to address this?

Reply:
I assume it is because `dependsOn()` method was added after `RemoteLoadBalancerCheckProviderFactory`. 
This is a "lazy provider" and is only instantiated when the REST endpoint is invoked.